walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,160 @@
+      optionalMeetingOptions = getMeetingRequest(allEventTimes, request.getDuration());
+      
+      if (!optionalMeetingOptions.isEmpty()) {
+        return optionalMeetingOptions;
+      }
+    }
+
+    return meetingOptions;
   }
+
+  private Collection<TimeRange> getMeetingRequest(ArrayList<TimeRange> eventTimes, Long duration) {
+    Collection<TimeRange> collection = new ArrayList<TimeRange>();
+    int currStart = TimeRange.START_OF_DAY;
+    int currEnd = TimeRange.END_OF_DAY;
+
+    // Create new time which starts at first opening and ends at  of request
+    newTime: 
+    for (TimeRange eventTime : eventTimes) {
+      currEnd = eventTime.start();
+      TimeRange proposeTime = TimeRange.fromStartEnd(currStart, currEnd, false);
+
+      // check that proposed time is longer than request dureation. Also ensures
+      // that the propsed duration is not negtive
+      if (checkDuration(proposeTime, duration)) {
+        // checks if proposed time overlaps other times and makes adjustments
+        // if duration is too short with new time, proposed time is nullified
+        proposeTime = checkProposedTime(proposeTime, duration, eventTimes);
+        if (proposeTime == null) {
+          currStart = eventTime.end();
+          continue newTime;
+        }
+
+        // if the proposed time is already within the collection don't add
+        for (TimeRange requestOption : collection) {
+          if (proposeTime.equals(requestOption)) {
+            currStart = eventTime.end();
+            continue;
+          }
+        }
+
+        collection.add(proposeTime);
+      }
+      currStart = eventTime.end();
+    }
+    //  add time gap for end of day, current start is on last 
+    TimeRange proposeTime = TimeRange.fromStartEnd(currStart, DAY_IN_MINS, false);
+    if (proposeTime.duration() > duration) {
+      collection.add(proposeTime);
+    }
+
+    return collection;
+  }
+
+  // returns a list of all event times based on attendee list, no duplicates.
+  private ArrayList<TimeRange> getAllEventTimes(Collection<String> attendees, Collection<Event> events) {
+    ArrayList<TimeRange> allEventTimes = new ArrayList<>();
+    Set<String> eventTitles = new HashSet<>();
+
+    for (String attendee : attendees) {
+      for (Event event : events) {
+        if (event.getAttendees().contains(attendee)) {
+          if (!eventTitles.contains(event.getTitle())) {
+            eventTitles.add(event.getTitle());
+            allEventTimes.add(event.getWhen());
+          }
+        }
+      }
+    }
+    return allEventTimes;
+  }
+
+  // Checks if proposed time based on attendence overlaps any other attendee events
+  // If there are overlaps then it returns the new proposed time 
+  private TimeRange checkProposedTime(TimeRange proposeTime, long duration, ArrayList<TimeRange> allEventTimes) {
+    for (TimeRange anEvent : allEventTimes) {
+      if (anEvent.equals(proposeTime)) {
+        continue;
```

Comment:
why don't we need to adjust the proposed time in this case?

Reply:
For some reason I thought proposeTime came from allEventTimes and so proposed would have been equal to one of the TimeRanges in allEventTimes. I realize now that's mot true at all. I've taken the if statement out and updated the comments